### DISCUSSION TITLE
Thread Safety in BLE communication

### DISCUSSION THREAD

**Discussion by nikamvd** [shinyorg/shiny#1184] [Q&A] [answered]
Are Shiny BLE api's thread safe by design?

For e.g Android BLE becomes unstable when we try to do BLE operations concurrently. We have to write custom queueing logic to ensure that no another write (or any ble operation) is performed when existing operation is in process. We should be ensuring that operation completed on the GattCallbacks.

Is this handled in Shiny by design?

**Comment by aritchie** [comment, marked as answer]
Yes - it is handled with an operation queue

**Reply by nikamvd** [reply]
@aritchie Thank you very much Allan for your quick response. Highly appreciated. Could you please point me in shiny BLE code where it is taken care? I tried but couldn't locate. I will appreciate if you can tell me Class name and methods where it is implemented. In my project, I have already implemented Bluetooth functionality. I have written it myself; completely custom separate on android (xamarin.android) and iOS (xamarin.iOS) and used it from xamarin.forms application. Now my android ble implementation is little unstable in concurrent situations. However iOS works good. If Shiny takes care of this by design, I can give thought to replace my custom implementation with Shiny instead of spending time correcting and fixing bugs in my custom solution.
